src/main/kotlin/org/wycliffeassociates/otter/jvm/persistence/repositories/WorkbookRepository.kt

```diff
@@ -0,0 +1,215 @@
+                    else -> constructResource(a, b)
+                }
+            }
+            .filterNotNull()
+    }
+
+    private fun workbookTake(modelTake: org.wycliffeassociates.otter.common.data.model.Take): Take {
+        // TODO: move into TakeMapper?
+        return Take(
+            name = modelTake.filename,
+            file = modelTake.path,
+            number = modelTake.number,
+            format = MimeType.WAV, // TODO
+            createdTimestamp = modelTake.timestamp,
+            deletedTimestamp = BehaviorRelay.createDefault(null) // TODO: need DB deleted field
+        )
+    }
+
+    private fun modelTake(workbookTake: Take): org.wycliffeassociates.otter.common.data.model.Take {
+        // TODO: move into TakeMapper?
+        return org.wycliffeassociates.otter.common.data.model.Take(
+            filename = workbookTake.file.name,
+            path = workbookTake.file,
+            number = workbookTake.number,
+            timestamp = workbookTake.createdTimestamp,
+            played = false,
+            markers = listOf()
+        )
+    }
+
+    private fun constructAssociatedAudio(content: Content): AssociatedAudio {
+        /** Map to recover model.Take objects from workbook.Take objects. */
+        val takeMap = WeakHashMap<Take, org.wycliffeassociates.otter.common.data.model.Take>()
+
+        /** Initial Takes read from the DB. */
+        val takesFromDb = takeRepository
+            .getByContent(content)
+            .flattenAsObservable { list -> list.sortedBy { it.number } }
+            .map { workbookTake(it) to it }
+
+        /** Relay to send Takes out to consumers, but also receive new Takes from UI. */
+        val takesRelay = ReplayRelay.create<Take>()
+        takesFromDb
+            // Record the mapping between data types.
+            .doOnNext { (wbTake, modelTake) -> takeMap[wbTake] = modelTake }
+            // Feed the initial list to takesRelay
+            .map { (wbTake, _) -> wbTake }
+            .subscribe(takesRelay)
+
+        /** When we receive new takes, update the map and write to the DB. */
+        val takesRelaySubscription = takesRelay
+            // Skip Takes we have seen before.
+            .filter { !takeMap.contains(it) }
+            // Make a Pair<workbook.Take, model.Take>
+            .map { it to modelTake(it) }
+            // Keep the takeMap current.
+            .doOnNext { (wbTake, modelTake) -> takeMap[wbTake] = modelTake }
+            // Insert the new take into the DB.
+            .subscribe { (_, modelTake) ->
+                takeRepository.insertForContent(modelTake, content)
+            }
+
+        /** The initial selected take, from the DB. */
+        val initialSelectedTake = content.selectedTake?.let { workbookTake(it) }
+
+        /** Relay to send selected-take updates out to consumers, but also receive updates from UI. */
+        val selectedTakeRelay = BehaviorRelay.createDefault(initialSelectedTake)
```

Comment:
I just discovered RxJava2 streams break on null elements.  So our nullable observables will need to be wrapped in Optionals.  Grr.

Reply:
Resolved in e898910